### PR TITLE
[docs]  Redesigned Selectors

### PR DESCRIPTION
**Pull request by AlexanderMoskovkin**

### CODE REVIEW

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 1]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+------ | -----
+`find(cssSelector)` | Finds all descendants of all nodes in the matching set and filters them either by CSS selector.
+`find(filterFn)` | Finds all descendants of all nodes in the matching set and filters them using `filterFn`. `filterFn` is a client function predicate that receives node.
+`parent()` | Finds all parents of all nodes in the matching set.
+`parent(index)` | Finds all parents of all nodes in the matching set and filters them by `index` (0 is closest).
+`parent(cssSelector)` | Finds all parents of all nodes in the matching set and filters them by `cssSelector`.
+`parent(filterFn)` | Finds all parents of all nodes in the matching set and filters them by `filterFn`. `filterFn` is a client function predicate that receives node.
+`child()` | Finds all child elements (not nodes) of all nodes in the matching set.
+`child(index)` | Finds all child elements (not nodes) of all nodes in the matching set and filters them by `index`.
+`child(cssSelector)` | Finds all parents of all nodes in the matching set and filters them by `cssSelector`.
```
Comment (helen-dikareva) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: oops, `parents`. The same below.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 2]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+const snapshot = await Selector('ul').find('label').parent('div.someClass').nth(4)();
+```
+
+------
+
+Note that you can add text and index filters in the selector chain:
+```js
+Selector('.container').parent(1).nth(0).find('.someClass').withText('yo!').child('span');
+```
+Finds second parents (parent of parent) of `.container` elements. Peeks first element in the matching set. Then in that element finds elements that matches '.someClass' selector. Filter them by text `yo!`. And in each filtered element searches for a child with tag name `span`.
```
Comment (helen-dikareva) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: May be '.someClass' should be highlighted like `.container`

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 3]
```diff
@@ -192,6 +300,7 @@ fixture `My fixture`
 
 test('My test', async t => {
     const button = await elementWithId('my-button');
+	const div    = await Selector('.myDiv')();
```
Comment (helen-dikareva) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: formating

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 4]
```diff
@@ -266,8 +373,9 @@ To do this, use the [visibilityCheck](selector-options.md#optionsvisibilitycheck
 
 ## One-Time Selection
 
-To create a selector and immediately execute it without saving it,
-use the `select` method of the [test controller](../test-code-structure.md#test-controller).
+> Important! ***deprecated*** *Will be removed in next releases. Use [selectors](link) and [property shorthands](link) to obtain values.*
```
Comment (helen-dikareva) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: We can add first link yet

**Review by helen-dikareva — CHANGES_REQUESTED** [review]
/r-

File: docs/articles/documentation/test-api/selecting-page-elements/selector-options.md [hunk 5]
```diff
@@ -2,25 +2,54 @@
...
 
 You can also overwrite the options you have specified before.
 
 * [Overwriting Options](#overwriting-options)
 
+## options.boundTestRun
+
+**Type**: Object
+
+If you need to call a selector from a Node.js callback, assign the current
```
Comment (lexkazakov) [inline docs/articles/documentation/test-api/selecting-page-elements/selector-options.md]: To call a selector from a Node.js callback, assign the current [test controller](../test-code-structure.md#test-controller) to the `boundTestRun` option.

File: docs/articles/documentation/test-api/selecting-page-elements/selector-options.md [hunk 6]
```diff
@@ -40,16 +69,8 @@ const gridCell = Selector((m, n) => gridRow(m).children[n], {
...
-
-**Type**: Object
-
-If you need to call a selector from a Node.js callback, assign the current
-[test controller](../test-code-structure.md#test-controller) to the `boundTestRun` option.
-
-For details, see [Calling Selectors from Node.js Callbacks](selectors.md#calling-selectors-from-nodejs-callbacks).
-
 ## options.text
+> Important! ***deprecated*** *Will be removed in next releases. Use [withText()](selectors.md#filter-multiple-dom-nodes) to filter elements by text.*
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selector-options.md]: ***Deprecated.*** *Use the [withText()](selectors.md#filter-multiple-dom-nodes) method to filter elements by text content or a regular expression that matches this content. The [options.text](#optionstext) option will be removed in future releases.*

File: docs/articles/documentation/test-api/selecting-page-elements/selector-options.md [hunk 7]
```diff
@@ -64,6 +85,7 @@ You can also use the [index](#optionsindex) option to identify a node by its ind
 For details, see [Initializers that Return Multiple Nodes](selectors.md#initializers-that-return-multiple-nodes).
 
 ## options.index
+> Important! ***deprecated*** *Will be removed in next releases. Use [nth()](selectors.md#filter-multiple-dom-nodes) to filter elements by its index.*
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selector-options.md]: ***Deprecated.*** *Use the [nth()](selectors.md#filter-multiple-dom-nodes) method to select a node by its index. The [options.index](#optionsindex) option will be removed in future releases.*

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 8]
```diff
@@ -74,7 +79,8 @@ You can initialize a selector with any of these objects.
     import { Selector } from 'testcafe';
 
     // A selector is created from a CSS selector string.
-    // This selector will return the first matching DOM node.
+    // This selector will return the first matching DOM node
+    // when you pass it to an action or an assertion or execute it.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: // This selector will return the first matching DOM node.
// You can execute this selector or pass it as a parameter to an action or assertion.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 9]
```diff
@@ -84,28 +90,26 @@ You can initialize a selector with any of these objects.
     import { Selector } from 'testcafe';
 
     // This selector is created from a function that returs all elements of a specified class.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: // This selector is created from a function that retur_n_s all visible elements of a specified class.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 10]
```diff
@@ -84,28 +90,26 @@ You can initialize a selector with any of these objects.
...
-    // The selector returns the third such element since its `index` option is set to 2.
-    const thirdElemWithClass = Selector(cl => document.getElementsByClassName(cl), {
-        index: 2
+    // The selector additionally requires the returned element to become visible.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: [Kazakov: Remove this code comment]

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 11]
```diff
@@ -84,28 +90,26 @@ You can initialize a selector with any of these objects.
...
+    const elemWithClass = Selector(cl => document.getElementsByClassName(cl), {
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: [Kazakov: Rename it to accent that selector returns *multiple* elements]

const elementsWithClass = Selector(cl => document.getElementsByClassName(cl), {

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 12]
```diff
@@ -84,28 +90,26 @@ You can initialize a selector with any of these objects.
...
+        visibilityCheck: true
     });
 
-    // This selector is created based on the previous selector and inherits its initializer,
-    // but overwrites the `index` parameter to return the fourth element.
-    const fourthElemWithClass = Selector(thirdElemWithClass, { index: 3 });
+    // This selector is created based on the previous selector and inherits
+    // its initializer, but overwrites the `visibilityCheck` parameter.
+    const fourthElemWithClass = Selector(thirdElemWithClass, { visibilityCheck: false });
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: [Kazakov: the previous selector was renamed, so rename the current one.]
const hiddenElementsWithClass = Selector(elementsWithClass, { visibilityCheck: false });

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 13]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
     });
     ```
 
-### Initializers that Return Multiple Nodes
+### Check Selector Matching Set Length
+
+Functions and CSS selector strings which initialize a selector can have one matching DOM element on the page, several or nothing.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: Functions and CSS selector strings that initialize a selector may return a single matching DOM element on the page, multiple elements or nothing.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 14]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+Selector provides the following properties to check this:
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: You can use the following Selector properties to check whether the matching elements exist or get a number of them.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 15]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
 
-The [text](selector-options.md#optionstext) option specifies text content of the node that should be returned
-or a regular expression that matches this content.
+Note that selector property getters are asynchronous.
 
-The [index](selector-options.md#optionsindex) option specifies the index of the node that should be returned.
+## Combined Selectors
 
-The following example demonstrates how to use the `text` and `index` options.
+Selector api provides methods that can be combined together, making flexible functional-style selector system.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: The selector API provides methods that can be combined together, thus providing you with a flexible functional-style selector mechanism.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 16]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+
+ 
+### Filter Multiple DOM Nodes
+
+If selector returns multiple DOM nodes you must filter them to select a single node that will eventually be returned by the selector.
+Selector provides methods to filter DOM nodes by its index and text:
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: Selector provides methods to filter DOM nodes by their index or text:

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 17]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+
+ 
+### Filter Multiple DOM Nodes
+
+If selector returns multiple DOM nodes you must filter them to select a single node that will eventually be returned by the selector.
+Selector provides methods to filter DOM nodes by its index and text: 
+
+Method | Type | Description 
+------ | ----- | -----
+`nth(index)` | Selector | Creates selector for `index` element in the matching set.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: Creates selector that return an element by its index in the matching set.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 18]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+`withText(text)` | Selector | Creates selector that filters matching set by given `text`.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: Creates selector that filters matching set by the specified text.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 19]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+`withText(re)` | Selector | Creates selector that filters matching set by given regular expression `re`.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: Creates selector that filters matching set using the specified regular expression.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 20]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
 });
+```
 
-test('The second disabled button is visible', async t => {
-    expect(await secondButton('disabled').visible).to.be.true;
+If all nodes are filtered out, the selector returns `null`.
+
+### Find Elements by DOM Hierarchy
+
+Selector provide methods to find elements by DOM hierarchy in jQuery-style.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: The selector API provides methods to find elements within a DOM hierarchy in jQuery style.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 21]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+
+Property | Description
+------ | -----
+`find(cssSelector)` | Finds all descendants of all nodes in the matching set and filters them either by CSS selector.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: Finds all descendants of all nodes in the matching set and filters them by `cssSelector`.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 22]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+`find(filterFn)` | Finds all descendants of all nodes in the matching set and filters them using `filterFn`. `filterFn` is a client function predicate that receives node.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: Finds all descendants of all nodes in the matching set and filters them using `filterFn`. `filterFn` is a client function predicate that receives a node.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 23]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+`parent()` | Finds all parents of all nodes in the matching set.
+`parent(index)` | Finds all parents of all nodes in the matching set and filters them by `index` (0 is closest).
+`parent(cssSelector)` | Finds all parents of all nodes in the matching set and filters them by `cssSelector`.
+`parent(filterFn)` | Finds all parents of all nodes in the matching set and filters them by `filterFn`. `filterFn` is a client function predicate that receives node.
+`child()` | Finds all child elements (not nodes) of all nodes in the matching set.
+`child(index)` | Finds all child elements (not nodes) of all nodes in the matching set and filters them by `index`.
+`child(cssSelector)` | Finds all child elements (not nodes) of all nodes in the matching set and filters them by `cssSelector`.
+`child(filterFn)` | Finds all child elements (not nodes) of all nodes in the matching set and filters them by `filterFn`. `filterFn` is a client function predicate that receives node.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md:223]: `filterFn` is a client function predicate that receives a node.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 24]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+`sibling()` | Finds all sibling  elements (not nodes) of all nodes in the matching set.
+`sibling(index)` | Finds all sibling  elements (not nodes) of all nodes in the matching set and filters them by `index`.
+`sibling(cssSelector)` | Finds all sibling  of all nodes in the matching set and filters them by `cssSelector`.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: Finds all sibling elements (not nodes) of all nodes in the matching set and filters them by `cssSelector`.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 25]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
+`sibling(filterFn)` | Finds all sibling  of all nodes in the matching set and filters them by `filterFn`. `filterFn` is a client function predicate that receives node.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: Finds all sibling elements (not nodes) of all nodes in the matching set and filters them by `filterFn`. `filterFn` is a client function predicate that receives a node.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 26]
```diff
@@ -138,45 +142,149 @@ You can initialize a selector with any of these objects.
...
 
-If both options are specified, nodes are first filtered by the `text` filter,
-then a node at the `index` position is selected from the remaining nodes.
+### Examples
 
-If all nodes are filtered out, the selector returns `null`.
+```js
+Selector('ul').find('label').parent('div.someClass')
+```
+Finds all `ul` elements on page. Then in each found `ul` element finds `label` elements. Then for each `label` element finds parent that matches  `div.someClass` selector.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: formatting

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 27]
```diff
@@ -192,9 +300,12 @@ fixture `My fixture`
 
 test('My test', async t => {
     const button = await elementWithId('my-button');
+	const div    = await Selector('.myDiv')(); 
 });
 ```
 
+Note that if selector initializer has several matching DOM nodes on the page the selector execution returns the first one from the set.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: Note that if a selector initializer has several matching DOM nodes on the page, the selector returns the first node from the matching set.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 28]
```diff
@@ -266,8 +375,9 @@ To do this, use the [visibilityCheck](selector-options.md#optionsvisibilitycheck
 
 ## One-Time Selection
 
-To create a selector and immediately execute it without saving it,
-use the `select` method of the [test controller](../test-code-structure.md#test-controller).
+> Important! ***deprecated*** *Will be removed in next releases. Use [selectors](#creating-selectors) instead.*
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: ***Deprecated*** *Use [selectors](#creating-selectors) instead. The `select` method of the [test controller](../test-code-structure.md#test-controller) will be removed in future releases.*

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 29]
```diff
@@ -346,6 +454,28 @@ Before executing an action, TestCafe waits for the target element to appear
 in the DOM and become visible. If this does not happen
 within the [selector timeout](#selector-timeout), the test fails.
 
+Note that if selector initializer has several matching DOM nodes on the page the action will be executed for the first one from the set.
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: Note that if a selector initializer has multiple matching DOM nodes on the page, the action will be executed only for the first node from the matching set.

File: docs/articles/documentation/test-api/selecting-page-elements/selectors.md [hunk 30]
```diff
@@ -346,6 +454,28 @@ Before executing an action, TestCafe waits for the target element to appear
...
+
+## Using Selectors to Define Assertion Actual Value
+
+You can check is a DOM node has an expected state by passing [snapshot API shorthand](#snapshot-api-shorthands) directly to [test actions](../assertions/index.md).
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selectors.md]: You can check whether a particular DOM node has an expected state by passing [snapshot API shorthand](#snapshot-api-shorthands) directly to [test actions](../assertions/index.md).

File: docs/articles/index.html [hunk 31]
```diff
@@ -67,12 +67,20 @@
                 <p>Additionally, TestCafe automatically generates source maps for easy debugging. To debug your test code, start a debugging session in an IDE that supports source maps.</p>
             </div>
         </div>
+		<div class="feature">
+            <span class="name">
+                Flexible Selectors System
+            </span>
+            <div class="description">
+                <p>You can access webpage elements in different ways (by css selector, text, DOM hierarchy or use any custom logic) with <a href="documentation/test-api/selecting-page-elements/selectors.html">selectors</a>. Это предоставлет широкие возможности для написания тестовых сценариев.</p>
```
Comment (alcazaco) [inline docs/articles/index.html]: [translate:] It provides you with rich capabilities for writing test scripts.

File: docs/articles/documentation/test-api/selecting-page-elements/selector-options.md [hunk 32]
```diff
@@ -2,25 +2,54 @@
...
+## options.visibilityCheck
+
+**Type**: Boolean
+
+`true` to additionally require the returned element to become visible within [options.timeout](#optionstimeout).
+
+**Default value**: `false`
+
 ## options.dependencies
+> Important! ***deprecated*** *Will be removed in next releases. Use [hierarchical selectors](selectors.md#find-elements-by-dom-hierarchy).*
```
Comment (alcazaco) [inline docs/articles/documentation/test-api/selecting-page-elements/selector-options.md]: ***Deprecated*** *Use [hierarchical selectors](selectors.md#find-elements-by-dom-hierarchy). The `option.dependencies` will be removed in future releases.*

**Pull request closed without merging**